borsh-derive-internal/src/enum_de.rs

```diff
@@ -70,9 +94,15 @@ pub fn enum_de(input: &ItemEnum, cratename: Ident) -> syn::Result<TokenStream2>
             }
             Fields::Unit => {}
         }
-        variant_arms.extend(quote! {
-            if variant_tag == #discriminant { #name::#variant_ident #variant_header } else
-        });
+        if use_discriminant {
+            variant_arms.extend(quote! {
+                if variant_tag == #discriminant { #name::#variant_ident #variant_header } else
```

Comment:
Ok, rust doesn't allow to use expressions at the spot, where a pattern is expected, so it should either be:
1. use `if-else_if-else` chain in both cases
2. or use `match` in both cases with [evalexpr crate](https://crates.io/crates/evalexpr), which works fine for integer literals in place of enum discriminants (but not other arbitrary expressions), similar to the following:
```diff
diff --git a/borsh-derive-internal/src/enum_discriminant_map.rs b/borsh-derive-internal/src/enum_discriminant_map.rs
index 9d7545a1..66f41ce0 100644
--- a/borsh-derive-internal/src/enum_discriminant_map.rs
+++ b/borsh-derive-internal/src/enum_discriminant_map.rs
@@ -1,24 +1,30 @@
 use std::collections::HashMap;
 
+use evalexpr::*;
+
 use proc_macro2::{Ident, TokenStream};
 use quote::quote;
 use syn::{punctuated::Punctuated, token::Comma, Variant};
 
 /// Calculates the discriminant that will be assigned by the compiler.
 /// See: https://doc.rust-lang.org/reference/items/enumerations.html#assigning-discriminant-values
-pub fn discriminant_map(variants: &Punctuated<Variant, Comma>) -> HashMap<Ident, TokenStream> {
+pub fn discriminant_map(variants: &Punctuated<Variant, Comma>) -> Result<HashMap<Ident, u8>, EvalexprError> {
     let mut map = HashMap::new();
 
     let mut next_discriminant_if_not_specified = quote! {0};
 
     for variant in variants {
         let this_discriminant = variant.discriminant.clone().map_or_else(
             || quote! { #next_discriminant_if_not_specified },
             |(_, e)| quote! { #e },
         );
         next_discriminant_if_not_specified = quote! { #this_discriminant + 1 };
-        map.insert(variant.ident.clone(), this_discriminant);
+
+        let value = eval(&this_discriminant.to_string())?;
+        let value_int = value.as_int()?;
+        // replace `as` cast with `u8::try_from(value_int).map_err(...)`
+        map.insert(variant.ident.clone(), value_int as u8);
     }
 
-    map
+    Ok(map)
 }
```